Mathlib/RingTheory/Bialgebra/TensorProduct.lean

```diff
@@ -0,0 +1,166 @@
+  rfl
+
+end Bialgebra.TensorProduct
+
+noncomputable instance TensorProduct.instBialgebra
+    {R A B : Type u} [CommRing R] [Ring A] [Ring B]
+    [Bialgebra R A] [Bialgebra R B] : Bialgebra R (A ⊗[R] B) := by
+  have hcounit := congr(DFunLike.coe $(Bialgebra.TensorProduct.counit_eq_algHom_toLinearMap R A B))
+  have hcomul := congr(DFunLike.coe $(Bialgebra.TensorProduct.comul_eq_algHom_toLinearMap R A B))
+  refine Bialgebra.mk' R (A ⊗[R] B) ?_ (fun {x y} => ?_) ?_ (fun {x y} => ?_) <;>
+  simp_all only [AlgHom.toLinearMap_apply] <;>
+  simp only [_root_.map_one, _root_.map_mul]
+
+namespace Bialgebra.TensorProduct
+
+variable {R A B C D : Type u} [CommRing R] [Ring A] [Ring B] [Ring C] [Ring D]
+    [Bialgebra R A] [Bialgebra R B] [Bialgebra R C] [Bialgebra R D]
+
+/-- The tensor product of two bialgebra morphisms as a bialgebra morphism. -/
+noncomputable def map (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    A ⊗[R] C →ₐc[R] B ⊗[R] D :=
+  { Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D),
+    Algebra.TensorProduct.map (f : A →ₐ[R] B) (g : C →ₐ[R] D) with }
+
+@[simp]
+theorem map_tmul (f : A →ₐc[R] B) (g : C →ₐc[R] D) (x : A) (y : C) :
+    map f g (x ⊗ₜ y) = f x ⊗ₜ g y :=
+  rfl
+
+@[simp]
+theorem map_toCoalgHom (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    map f g = Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D) := rfl
+
+variable (R A B C)
+
+/-- The associator for tensor products of R-bialgebras, as a bialgebra equivalence. -/
+protected noncomputable def assoc :
+    (A ⊗[R] B) ⊗[R] C ≃ₐc[R] A ⊗[R] (B ⊗[R] C) :=
+  { Coalgebra.TensorProduct.assoc R A B C, Algebra.TensorProduct.assoc R A B C with }
+
+variable {R A B C}
+
+@[simp]
+theorem assoc_tmul (x : A) (y : B) (z : C) :
+    Bialgebra.TensorProduct.assoc R A B C ((x ⊗ₜ y) ⊗ₜ z) = x ⊗ₜ (y ⊗ₜ z) :=
+  rfl
+
+@[simp]
+theorem assoc_symm_tmul (x : A) (y : B) (z : C) :
+    (Coalgebra.TensorProduct.assoc R A B C).symm (x ⊗ₜ (y ⊗ₜ z)) = (x ⊗ₜ y) ⊗ₜ z :=
+  rfl
+
+@[simp]
+theorem assoc_toCoalgEquiv :
+    (Bialgebra.TensorProduct.assoc R A B C : _ ≃ₗc[R] _) =
+    Coalgebra.TensorProduct.assoc R A B C := rfl
+
+variable (R A)
+
+/-- The base ring is a left identity for the tensor product of bialgebras, up to
+bialgebra equivalence. -/
+protected noncomputable def lid : R ⊗[R] A ≃ₐc[R] A :=
+  { Coalgebra.TensorProduct.lid R A, Algebra.TensorProduct.lid R A with }
+
+variable {R A}
+
+@[simp]
+theorem lid_toCoalgEquiv :
+    (Bialgebra.TensorProduct.lid R A : R ⊗[R] A ≃ₗc[R] A) = Coalgebra.TensorProduct.lid R A := rfl
+
+@[simp]
+theorem lid_tmul (r : R) (a : A) : Bialgebra.TensorProduct.lid R A (r ⊗ₜ a) = r • a := rfl
+
+@[simp]
+theorem lid_symm_apply (a : A) : (Bialgebra.TensorProduct.lid R A).symm a = 1 ⊗ₜ a := rfl
+
+theorem coalgebraRid_eq_algebraRid_apply (x) :
+    Coalgebra.TensorProduct.rid R A x = Algebra.TensorProduct.rid R R A x :=
+  congr($((TensorProduct.AlgebraTensorModule.rid_eq_rid R A).symm) x)
+
+variable (R A)
```

Comment:
Instead of making them explicit and immediately implicit again, you can just do `variable (R A) in`

Reply:
thanks, done